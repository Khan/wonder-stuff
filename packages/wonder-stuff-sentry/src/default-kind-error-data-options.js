// @flow
import type {KindErrorDataOptions} from "./types";

/**
 * Default values for the `KindErrorDataOptions` type.
 */
export const DefaultKindErrorDataOptions: KindErrorDataOptions = {
    kindTagName: "kind",
    groupByTagName: "group_by_message",
    concatenatedMessageTagName: "concatenated_message",
    causalErrorContextPrefix: "Source Error - ",
};
