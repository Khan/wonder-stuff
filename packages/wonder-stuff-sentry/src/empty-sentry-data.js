// @flow
import type {SentryData} from "./types";

/**
 * An empty sentry data object.
 */
export const EmptySentryData: $ReadOnly<SentryData> = Object.freeze({
    tags: {},
    contexts: {},
    fingerprint: [],
});
