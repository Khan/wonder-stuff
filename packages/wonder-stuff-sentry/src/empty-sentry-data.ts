import type {SentryData} from "./types";

/**
 * An empty sentry data object.
 */
export const EmptySentryData: Readonly<SentryData> = Object.freeze({
    tags: {},
    contexts: {},
    fingerprint: [],
});
