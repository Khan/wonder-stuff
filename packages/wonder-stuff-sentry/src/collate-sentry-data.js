// @flow
import {
    errorsFromError,
    Order,
    getKindFromError,
} from "@khanacademy/wonder-stuff-core";
import {EmptySentryData} from "./empty-sentry-data.js";
import {sentryDataReducer} from "./sentry-data-reducer.js";
import {KindSentryError} from "./kind-sentry-error.js";
import {normalizeSentryData} from "./normalize-sentry-data.js";
import {truncateTagValue} from "./truncate-tag-value.js";
import type {SentryData, KindErrorDataOptions} from "./types.js";

/**
 * Collate sentry data from a given error and its causal errors into one object.
 *
 * Parses the error and all causal errors, merging any SentryData found to
 * produce a single SentryData object.
 *
 * For `tags` and `contexts`, existing values of inner (earlier) errors are
 * overwritten by outer (later) errors. For `fingerprints`, the arrays are
 * concatenated together.
 *
 * Each causal error is added as a context containing its original stack,
 * sentry data, and message.
 *
 * Finally, tags for the error kind, grouping, and message concatenation are
 * added.
 *
 * @param {Error} error The error to be collated into sentry data.
 * @returns {$ReadOnly<SentryData>} A single SentryData object combining
 * all the SentryData from this and causal errors it contains.
 */
export const collateSentryData = (
    options: KindErrorDataOptions,
    error: Error,
): $ReadOnly<SentryData> => {
    // First, get all the errors in the stack.
    const errors = errorsFromError(error, Order.ConsequenceFirst);
    const consquenceAndCauses: $ReadOnlyArray<Error> = Array.from(errors);

    // Now, we need to collate this stack as follows:
    // 1. Any sentryData they hold has to be merged into the collatedData from
    //    the cause to the consquence, so that the more recent consequences
    //    take precedent.
    // 2. Errors, regardless of whether they have sentryData, need to be
    //    given their own context in the data.
    const rawCollatedData: $ReadOnly<SentryData> =
        consquenceAndCauses.reduceRight(
            (acc, value, index) =>
                sentryDataReducer(options, acc, value, index),
            EmptySentryData,
        );

    // Before we add our additional tags, we need to ensure that the data is
    // valid and normalized.
    const collatedData = normalizeSentryData(options, rawCollatedData);

    // Finally, add the tags for kind, consequence, and group by.
    const {groupByTagName, concatenatedMessageTagName, kindTagName} = options;
    const kindTag = getKindFromError(error);

    /**
     * We always set the `kind` tag.
     * We set the tag `concatenated_message` to the concatenation of all
     * the wrapped errors' messages, to allow for wildcard searching on a
     * single Sentry Tag (once available in our plan):
     *     e.g., `concatenated_message:'*partial text of inner error*'`
     *
     * We set the tag `group_by_message`, as we have a toggleable
     * server-side rule that changes our grouping to use this value.
     *     `tags.group_by_message:"*" -> {{ tags.group_by_message }}`
     */
    const groupByTag = error.message.trim().split("\n")[0].trim();
    const concatenatedMessageTag =
        error instanceof KindSentryError
            ? error.message.replace(/[\n]+[\t]*/g, " ")
            : error.message.split("\n")[0].trim();

    const tags = {
        ...collatedData.tags,
        [kindTagName]: truncateTagValue(kindTag),
        [concatenatedMessageTagName]: truncateTagValue(concatenatedMessageTag),
    };
    if (groupByTag.length) {
        tags[groupByTagName] = truncateTagValue(groupByTag);
    }

    return {
        tags,
        contexts: collatedData.contexts,
        fingerprint: collatedData.fingerprint,
    };
};
