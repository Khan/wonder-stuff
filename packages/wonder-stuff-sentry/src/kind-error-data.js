// @flow
import type {
    SentryEvent,
    SentryEventHint,
    SentryEventProcessor,
    SentryHub,
    SentryIntegration,
    KindErrorDataOptions,
} from "./types.js";

import {collateSentryData} from "./collate-sentry-data.js";
import {DefaultKindErrorDataOptions} from "./default-kind-error-data-options.js";

export class KindErrorData implements SentryIntegration {
    static id: string = "KindErrorData";
    name: string = KindErrorData.id;
    +_options: KindErrorDataOptions;

    constructor(options: KindErrorDataOptions = DefaultKindErrorDataOptions) {
        this._options = options;
    }

    setupOnce(
        addGlobalEventProcessor: (callback: SentryEventProcessor) => void,
        getCurrentHub: () => SentryHub,
    ): void {
        addGlobalEventProcessor(
            (event: SentryEvent, hint?: SentryEventHint) => {
                const self = getCurrentHub().getIntegration(KindErrorData);
                if (!self) {
                    return event;
                }
                return self.enhanceEventWithErrorData(event, hint);
            },
        );
    }

    /**
     * Attaches extracted information from the Error object to extra field in the Event
     */
    enhanceEventWithErrorData(
        event: SentryEvent,
        hint?: SentryEventHint,
    ): SentryEvent {
        const maybeError = hint?.originalException;
        // We only enhance events of type error.
        if (!(maybeError instanceof Error)) {
            return event;
        }

        // Collate the data we want to collect.
        const {tags, contexts, fingerprint} = collateSentryData(
            this._options,
            maybeError,
        );

        // Now that we have data, we need to attach it to the event.
        /**
         * Tags help categorize things.
         */
        event.tags = {
            ...event.tags,
            ...tags,
        };

        /**
         * Each context creates a new headed section in the sentry report.
         * Useful for grouping specific context together.
         *
         * We now output all the contexts to the scope.
         */
        event.contexts = {
            ...event.contexts,
            ...contexts,
        };

        /**
         * Fingerprint helps group like messages that otherwise would not
         * get grouped.
         */
        event.fingerprint = [...(event.fingerprint ?? []), ...fingerprint];

        return event;
    }
}
