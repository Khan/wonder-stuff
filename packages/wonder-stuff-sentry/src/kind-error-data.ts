import {Errors} from "@khanacademy/wonder-stuff-core";
import type {
    SentryEvent,
    SentryEventHint,
    SentryEventProcessor,
    SentryHub,
    SentryIntegration,
    KindErrorDataOptions,
} from "./types";

import {collateSentryData} from "./collate-sentry-data";
import {DefaultKindErrorDataOptions} from "./default-kind-error-data-options";
import {isTagKeyValid} from "./is-tag-key-valid";
import {KindSentryError} from "./kind-sentry-error";

type InvalidTags = {
    invalidKindTag?: string;
    invalidGroupByTag?: string;
    invalidConcatenatedMessageTag?: string;
};

export class KindErrorData implements SentryIntegration {
    static id: string = "KindErrorData";
    name: string = KindErrorData.id;
    readonly _options: KindErrorDataOptions;

    constructor(options: Partial<KindErrorDataOptions> = Object.freeze({})) {
        this._options = {
            ...DefaultKindErrorDataOptions,
            ...options,
        };

        if (process.env.NODE_ENV !== "production") {
            // Let's make sure we got valid options.
            const invalidTagNames: InvalidTags = {};
            if (!isTagKeyValid(this._options.kindTagName)) {
                invalidTagNames.invalidKindTag = this._options.kindTagName;
            }
            if (!isTagKeyValid(this._options.groupByTagName)) {
                invalidTagNames.invalidGroupByTag =
                    this._options.groupByTagName;
            }
            if (!isTagKeyValid(this._options.concatenatedMessageTagName)) {
                invalidTagNames.invalidConcatenatedMessageTag =
                    this._options.concatenatedMessageTagName;
            }
            if (Object.keys(invalidTagNames).length) {
                throw new KindSentryError(
                    "Invalid options",
                    Errors.InvalidInput,
                    {
                        sentryData: {
                            contexts: {
                                invalidTagNames: {
                                    ...invalidTagNames,
                                },
                            },
                        },
                    },
                );
            }
        }
    }

    setupOnce(
        addGlobalEventProcessor: (callback: SentryEventProcessor) => void,
        getCurrentHub: () => SentryHub,
    ): void {
        addGlobalEventProcessor(
            (event: SentryEvent, hint?: SentryEventHint) => {
                const self = getCurrentHub().getIntegration(KindErrorData);
                if (!self) {
                    return event;
                }
                return self.enhanceEventWithErrorData(event, hint);
            },
        );
    }

    /**
     * Attaches extracted information from the Error object to extra field in the Event
     */
    enhanceEventWithErrorData(
        event: SentryEvent,
        hint?: SentryEventHint,
    ): SentryEvent {
        const maybeError = hint?.originalException;
        // We only enhance events of type error.
        if (!(maybeError instanceof Error)) {
            return event;
        }

        // Collate the data we want to collect.
        const {tags, contexts, fingerprint} = collateSentryData(
            this._options,
            maybeError,
        );

        // Now that we have data, we need to attach it to the event.
        /**
         * Tags help categorize things.
         */
        event.tags = {
            ...event.tags,
            ...tags,
        };

        /**
         * Each context creates a new headed section in the sentry report.
         * Useful for grouping specific context together.
         *
         * We now output all the contexts to the scope.
         */
        const updatedContexts = {
            ...event.contexts,
            ...contexts,
        } as const;
        /**
         * NOTE: If you don't see Sentry serializing the right depth in your
         * contexts, increase the `normalizeDepth` option of the Sentry
         * configuration; it defaults to 3, which is not always enough.
         */
        event.contexts = updatedContexts;

        /**
         * Fingerprint helps group like messages that otherwise would not
         * get grouped.
         */
        event.fingerprint = [...(event.fingerprint ?? []), ...fingerprint];

        return event;
    }
}
