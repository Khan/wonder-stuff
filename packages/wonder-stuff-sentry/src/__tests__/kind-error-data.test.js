// @flow
import * as CollateSentryData from "../collate-sentry-data";
import {DefaultKindErrorDataOptions} from "../default-kind-error-data-options";
import {EmptySentryData} from "../empty-sentry-data";
import {KindErrorData} from "../kind-error-data";

jest.mock("../collate-sentry-data");

describe("KindErrorData", () => {
    const NODE_ENV = process.env.NODE_ENV;
    afterEach(() => {
        if (NODE_ENV == null) {
            delete process.env.NODE_ENV;
        } else {
            process.env.NODE_ENV = NODE_ENV;
        }
    });

    const INVALID_TAG_NAME = Array(33).fill("a").join("");
    describe("#constructor", () => {
        describe("when not built for production", () => {
            it.each([
                {kindTagName: INVALID_TAG_NAME},
                {groupByTagName: INVALID_TAG_NAME},
                {concatenatedMessageTagName: INVALID_TAG_NAME},
                {
                    kindTagName: `${INVALID_TAG_NAME}-kind`,
                    groupByTagName: `${INVALID_TAG_NAME}-group-by`,
                    concatenatedMessageTagName: `${INVALID_TAG_NAME}-concatenated-message`,
                },
            ])("should throw if options are invalid (%s)", (badOptions) => {
                // Arrange

                // Act
                const act = () => new KindErrorData(badOptions);

                // Assert
                expect(act).toThrowErrorMatchingSnapshot();
            });
        });

        describe("when built for production", () => {
            beforeEach(() => {
                process.env.NODE_ENV = "production";
            });

            it.each([
                {kindTagName: INVALID_TAG_NAME},
                {groupByTagName: INVALID_TAG_NAME},
                {concatenatedMessageTagName: INVALID_TAG_NAME},
                {
                    kindTagName: `${INVALID_TAG_NAME}-kind`,
                    groupByTagName: `${INVALID_TAG_NAME}-group-by`,
                    concatenatedMessageTagName: `${INVALID_TAG_NAME}-concatenated-message`,
                },
            ])("should not throw if options are invalid (%s)", (badOptions) => {
                // Arrange

                // Act
                const act = () => new KindErrorData(badOptions);

                // Assert
                expect(act).not.toThrowError("Invalid options");
            });
        });

        it("should include information when throwing about invalid options", async () => {
            // Arrange
            const badOptions = {
                kindTagName: `${INVALID_TAG_NAME}-kind`,
                groupByTagName: `${INVALID_TAG_NAME}-group-by`,
                concatenatedMessageTagName: `${INVALID_TAG_NAME}-concatenated-message`,
            };

            // Act
            const act = new Promise((resolve, reject) => {
                try {
                    // Should not resolve!
                    resolve(new KindErrorData(badOptions));
                } catch (e) {
                    reject(e);
                }
            });

            // Assert
            await expect(act).rejects.toHaveProperty("sentryData", {
                tags: {},
                fingerprint: [],
                contexts: {
                    invalidTagNames: {
                        invalidKindTag: badOptions.kindTagName,
                        invalidGroupByTag: badOptions.groupByTagName,
                        invalidConcatenatedMessageTag:
                            badOptions.concatenatedMessageTagName,
                    },
                },
            });
        });
    });

    describe("#setupOnce", () => {
        it("should register a global event processor", () => {
            // Arrange
            const addGlobalEventProcessorMock = jest.fn();
            const underTest = new KindErrorData();

            // Act
            underTest.setupOnce(addGlobalEventProcessorMock, jest.fn());

            // Assert
            expect(addGlobalEventProcessorMock).toHaveBeenCalledWith(
                expect.any(Function),
            );
        });

        describe("registered event processor", () => {
            it("should return the event if the integration is not returned from the hub", () => {
                // Arrange
                const addGlobalEventProcessorMock = jest.fn();
                const getHub = () => ({
                    getIntegration: () => undefined,
                });
                const underTest = new KindErrorData();
                underTest.setupOnce(addGlobalEventProcessorMock, getHub);
                const registeredProcessor =
                    addGlobalEventProcessorMock.mock.calls[0][0];
                const event = {
                    kind: "event",
                    event_id: "event-id",
                };

                // Act
                const result = registeredProcessor(event);

                // Assert
                expect(result).toBe(event);
            });

            it("should call enhanceEventWithErrorData", () => {
                // Arrange
                const underTest = new KindErrorData();
                const addGlobalEventProcessorMock = jest.fn();
                const getHub = () => ({
                    // $FlowIgnore[incompatible-call]
                    getIntegration: jest.fn().mockReturnValue(underTest),
                });
                const enhanceSpy = jest
                    .spyOn(underTest, "enhanceEventWithErrorData")
                    .mockReturnValue("ENHANCED");
                // $FlowIgnore[escaped-generic]
                underTest.setupOnce(addGlobalEventProcessorMock, getHub);
                const registeredProcessor =
                    addGlobalEventProcessorMock.mock.calls[0][0];
                const event = {
                    kind: "event",
                    event_id: "event-id",
                };
                const hint = {
                    event_id: "event-id",
                };

                // Act
                const result = registeredProcessor(event, hint);

                // Assert
                expect(enhanceSpy).toHaveBeenCalledWith(event, hint);
                expect(result).toBe("ENHANCED");
            });
        });
    });

    describe("#enhanceEventWithErrorData", () => {
        it.each([
            null,
            undefined,
            {
                kind: "not-an-error",
            },
        ])(
            "should return the event if the original exception (%s) is not an Error",
            (originalException) => {
                // Arrange
                const underTest = new KindErrorData();
                const event = {
                    kind: "event",
                    event_id: "event-id",
                };
                const hint = {
                    event_id: "event-id",
                    originalException,
                };

                // Act
                const result = underTest.enhanceEventWithErrorData(event, hint);

                // Assert
                expect(result).toBe(event);
            },
        );

        it("should return the event if there is no hint", () => {
            // Arrange
            const underTest = new KindErrorData();
            const event = {
                kind: "event",
                event_id: "event-id",
            };

            // Act
            const result = underTest.enhanceEventWithErrorData(event);

            // Assert
            expect(result).toBe(event);
        });

        it("should call collateSentryData with the original exception and constructed options", () => {
            // Arrange
            const options = {
                ...DefaultKindErrorDataOptions,
                groupByTagName: "GROUP",
                kindTagName: "KIND",
            };
            const underTest = new KindErrorData(options);
            const event = {
                kind: "event",
                event_id: "event-id",
            };
            const hint = {
                event_id: "event-id",
                originalException: new Error("original-exception"),
            };
            const collateSentryDataSpy = jest
                .spyOn(CollateSentryData, "collateSentryData")
                .mockReturnValue(EmptySentryData);

            // Act
            underTest.enhanceEventWithErrorData(event, hint);

            // Assert
            expect(collateSentryDataSpy).toHaveBeenCalledWith(
                options,
                hint.originalException,
            );
        });

        it("should combine the tags, contexts and fingerprints from collateSentryData with the event data given", () => {
            // Arrange
            const options = {
                ...DefaultKindErrorDataOptions,
                groupByTagName: "GROUP",
                kindTagName: "KIND",
            };
            const underTest = new KindErrorData(options);
            const event = {
                kind: "event",
                event_id: "event-id",
                tags: {
                    tag1: "originalValue1",
                    tag2: "originalValue2",
                },
                contexts: {
                    context1: {
                        key1: "originalValue1",
                    },
                    context2: {
                        key2: "originalValue2",
                    },
                },
                fingerprint: ["originalFingerprint"],
            };
            const hint = {
                event_id: "event-id",
                originalException: new Error("original-exception"),
            };
            jest.spyOn(CollateSentryData, "collateSentryData").mockReturnValue({
                tags: {
                    tag1: "tag1-value",
                },
                contexts: {
                    context1: {
                        key1: "value1",
                    },
                },
                fingerprint: ["fingerprint1", "fingerprint2"],
            });

            // Act
            const result = underTest.enhanceEventWithErrorData(event, hint);

            // Assert
            expect(result).toStrictEqual({
                ...event,
                tags: {
                    tag1: "tag1-value",
                    tag2: "originalValue2",
                },
                contexts: {
                    context1: {
                        key1: "value1",
                    },
                    context2: {
                        key2: "originalValue2",
                    },
                },
                fingerprint: [
                    "originalFingerprint",
                    "fingerprint1",
                    "fingerprint2",
                ],
            });
        });
    });
});
