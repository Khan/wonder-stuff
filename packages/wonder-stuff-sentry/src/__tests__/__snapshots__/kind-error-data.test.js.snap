// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KindErrorData #constructor when not built for production should throw if options are invalid ({
  kindTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-kind',
  groupByTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-group-by',
  concatenatedMessageTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-concatenated-message'
}) 1`] = `"Invalid options"`;

exports[`KindErrorData #constructor when not built for production should throw if options are invalid ({ concatenatedMessageTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' }) 1`] = `"Invalid options"`;

exports[`KindErrorData #constructor when not built for production should throw if options are invalid ({ groupByTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' }) 1`] = `"Invalid options"`;

exports[`KindErrorData #constructor when not built for production should throw if options are invalid ({ kindTagName: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' }) 1`] = `"Invalid options"`;
